File: src/5-Problem12.cpp
#include <iostream>
#include <fstream>
#include <string>
using namespace std;

int main()
{

  ifstream infile;
  int str;
  int l = 0;
  int lahan[100][100];
  bool lahancek[100][100];

  
  infile.open("contour.txt");
  if (!infile) {
      cout << "Unable to open file";
  }
  int j = -1;
  while (infile >> str){
    if (j == -1) {
      l = str;
      //cout << l << endl;
    } else {
      lahan[j][0] = str;
      lahancek[j][0] = false;
      for (int i = 1; i < l; i++) {
        infile >> str;
        lahan[j][i] = str;
        lahancek[j][i] = false;
      }
    }
    j++;
  }
  cout << "file loaded!" << endl;

  infile.close();

  int biggest = 0;

  for (int i = 0; i < l; i++) {
    for (j = 0; j < l; j++) {
      if (!lahancek[i][j]) {
        int tempNum = lahan[i][j];
        //cout << tempNum <<  " i: " << i  << " j: " << j << endl;
        int tempSize;

        int iteraTemp = 1;
        bool isSquare = true;
        while (isSquare) {
          tempSize = iteraTemp - 1;
          for (int iTemp = i; iTemp < i + iteraTemp; iTemp++) {
            if (!(tempNum == lahan[iTemp][j + iteraTemp - 1])) {
              isSquare = false;
              break;
            }
          }
          for (int iTemp = j; iTemp < j + iteraTemp; iTemp++) {
            if (!(tempNum == lahan[i + iteraTemp - 1][iTemp])) {
              isSquare = false;
              break;
            }
          }
          if (isSquare) {
            for (int iTemp = i; iTemp < i + iteraTemp; iTemp++) {
              lahancek[iTemp][j + iteraTemp - 1] = true;
            }
            for (int iTemp = i; iTemp < j + iteraTemp; iTemp++) {
              lahancek[i + iteraTemp - 1][iTemp] = true;
            }

            iteraTemp++;

          }
        }

        if (tempSize > biggest) {
          biggest = tempSize;
        }

      }
    }
  }

  cout << "Luas maksimum Krasti Krab yang dapat dibangun adalah " << biggest*biggest << endl;

  return 0;
}

File: src/5-Problem12.py
def is_matrix_valid(list1):
    list2 = [list1[i][j] for i in range(len(list1)) for j in range(len(list1))]
    list_set = set(list2)
    unique_list = list(list_set)
    return len(unique_list) == 1

if __name__ == "__main__":
    file_name = "contour.txt"

    try:
        f_input = open(file_name, "r")
        print("File loaded!")
    except IOError:
        print(file_name, "not found!")
    else:
        n = int(f_input.readline())

        map = []

        for _ in range(n):
            temp = f_input.readline().split()
            map.append(temp)

        #print(map)
        #print(len(map))

        i = 0
        j = 0
        # Pemeriksaan luas 1x1 tidak dilakukan
        k = 2
        offset = n - 1
        max_size = 1
        location = [0, 0]

        while i < offset and j < offset:
            #print("i = {}, j = {}".format(i,j))
            while i + k <= n and j + k <= n:
                sub_matrix_slice_1 = slice(i, i + k)
                sub_matrix_slice_j = slice(j, j + k)
                map_sliced = [map[i2][sub_matrix_slice_1] for i2 in range(len(map))[sub_matrix_slice_j]]

                #print(map_sliced)
                if is_matrix_valid(map_sliced):
                    if k ** 2 > max_size :
                        max_size = k ** 2
                        location[0] = i
                        location[1] = j
                        #print("New size {} at {}".format(max_size, location))
                else:
                    # Exit k loop
                    #print("Next Loop")
                    break

                k += 1

            k = 2
            i += 1
            if i == offset:
                i = 0
                j += 1

        print("Luas maksimum Krasti Krab yang dapat dibangun adalah {} dengan lokasi di {} ".format(max_size, location))
        f_input.close()